src/Discord/Internal/Types/ApplicationCommands.hs

```diff
@@ -0,0 +1,396 @@
+              ("choices", toJSON <$> optionChoices),
+              ("options", toJSON <$> optionOptions),
+              ("channel_types", toJSON <$> optionChannelTypes),
+              ("min_val", toJSON <$> optionMinVal),
+              ("max_val", toJSON <$> optionMaxVal),
+              ("autocomplete", toJSON <$> optionAutocomplete)
+            ]
+      ]
+
+instance FromJSON ApplicationCommandOption where
+  parseJSON =
+    withObject
+      "ApplicationCommandOption"
+      ( \v ->
+          ApplicationCommandOption
+            <$> v .: "type"
+            <*> v .: "name"
+            <*> v .: "description"
+            <*> v .:? "required"
+            <*> v .:? "choices"
+            <*> v .:? "options"
+            <*> v .:? "channel_types"
+            <*> v .:? "min_val"
+            <*> v .:? "max_val"
+            <*> v .:? "autocomplete"
+      )
+
+-- | What type of command option. Can represent a wide variety of types, so
+-- please check out the documentation below.
+--
+-- https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-type
+data ApplicationCommandOptionType
+  = -- | A subcommand. It can take further options, excluding sub commands and
+    -- sub command groups.
+    SUB_COMMAND
+  | -- | A subcommand group. It can take further options, excluding sub command
+    -- groups.
+    SUB_COMMAND_GROUP
+  | -- | Can typically be provided with default values.
+    STRING
+  | -- | Can typically be provided with default values, and possibly with
+    -- minimum and maximum values.
+    INTEGER
+  | BOOLEAN
+  | USER
+  | -- | Can be limited in the types of the channel allowed.
+    CHANNEL
+  | ROLE
+  | -- | Users and roles.
+    MENTIONABLE
+  | -- | Can typically be provided with default values, and possibly with
+    -- minimum and maximum values. Represents a double.
+    NUMBER
+  deriving (Show, Read, Data, Eq)
+
+instance Enum ApplicationCommandOptionType where
+  fromEnum SUB_COMMAND = 1
+  fromEnum SUB_COMMAND_GROUP = 2
+  fromEnum STRING = 3
+  fromEnum INTEGER = 4
+  fromEnum BOOLEAN = 5
+  fromEnum USER = 6
+  fromEnum CHANNEL = 7
+  fromEnum ROLE = 8
+  fromEnum MENTIONABLE = 9
+  fromEnum NUMBER = 10
+  toEnum a = fromJust $ lookup a table
+    where
+      table = makeTable SUB_COMMAND
+
+instance ToJSON ApplicationCommandOptionType where
+  toJSON = toJSON . fromEnum
+
+instance FromJSON ApplicationCommandOptionType where
+  parseJSON = withScientific "ApplicationCommandOptionType" (return . toEnum . round)
+
+-- | Utility data type to store integers or doubles.
+data IntDouble = IDI Int | IDD Double
+  deriving (Show, Read, Eq)
+
+instance ToJSON IntDouble where
+  toJSON (IDI s) = toJSON s
+  toJSON (IDD i) = toJSON i
+
+instance FromJSON IntDouble where
+  parseJSON v = (IDI <$> parseJSON v) <|> (IDD <$> parseJSON v)
+
+-- | Utility data type to store strings or number types.
+data StringNumber = SNS String | SNN IntDouble
```

Comment:
Hmm. I'd prefer the ADT name prefixed, but that would be an ugly name (`StringNumberString`?) Hmm.

Reply:
I can try to come up with an alternative name ig